shared/extension/ModuleTest/EmissionCatcher.hpp

```diff
@@ -0,0 +1,70 @@
+/*
+ * This file is part of the NUbots Codebase.
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2021 NUbots <nubots@nubots.net>
+ */
+
+#ifndef EXTENSION_MODULETEST_EMISSIONCATCHER_HPP
+#define EXTENSION_MODULETEST_EMISSIONCATCHER_HPP
+
+#include <catch.hpp>
+#include <nuclear>
+
+namespace extension::moduletest {
+
+    template <typename MessageType>
+    struct EmissionBind {
+        EmissionBind() = delete;
+        explicit EmissionBind(std::shared_ptr<MessageType> message_) : message(message_){};
+        std::shared_ptr<MessageType> message;
+    };
+
+    class EmissionCatcher : public NUClear::Reactor {
+    public:
+        template <typename MessageType>
+        explicit EmissionCatcher(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+            on<Trigger<extension::moduletest::EmissionBind<MessageType>>>().then(
+                // TODO: This doesn't work, because this wants me to know MessageType when I construct the
```

Comment:
It's kinda clunky but would it work if you did an `on<Always>` and then check if the type matches inside `EmissionCatcher` and if not return?

Reply:
I think I would have to know the return type to specify the "with"